MP4 support for HTTP push.

j0sh opened the pull request:
**What does this pull request do? Explain your changes. (required)**
<!-- A clear and concise description of what this pull request does. -->

This implements most of https://github.com/livepeer/go-livepeer/issues/1418 . The exception is that it does not implement output configuration via webhook. The output can be set to MP4 via the HTTP push file extension. This can be quickly tested on the broadcaster via:

```
ffmpeg -re -i input -c:a copy -c:v copy -f hls -hls_segment_filename "http://localhost:8935/live/mp4/%d.mp4" -
```

and the results inspected via `curl` the resulting playlist(s) at http:localhost:8935/live/mp4.m3u8 . ffplay of individual segments will work as well.

**Specific updates (required)**

- server: Take MP4s as input for HTTP push …
Most of the changes are in the following areas:

  * Adds a format -> extension lookup function
  * Adds a extension -> format inverse lookup function
  * Adds a format -> mime type lookup function
  * Protocol Buffers changes, and conversions to/from VideoProfile
  * Slotting in the lookups and error handling where appropriate
  * Lots and lots of tests.

The extensive testing and strict error handling is both to codify
intended behavior and to ensure misuse resistance to further changes. https://github.com/livepeer/go-livepeer/pull/1429/commits/949a06320544d326c25766f1da7c162c181a1e5e

- The file extension of temp files mostly doesn't matter for us anymore, so renamed the temp files we generate in order to avoid future confusion about the file type. https://github.com/livepeer/go-livepeer/pull/1429/commit/c23fc6079fd76d9c7564f3345df7695e4274b6bf

- Added documentation https://github.com/livepeer/go-livepeer/pull/1429/commits/39598171b72337b4f85b7aa3b43fcbecada6f153

- Updated LPMS to https://github.com/livepeer/lpms/pull/176 (needs to be merged prior to this ; any changes there will be updated here in go.mod)


**How did you test each of these updates (required)**
- Unit testing. Tests are the bulk of the additions here.
- Manual testing / playback with the following simple scenarios (B -> OT, local storage on both, played back via ffplay):
  * RTMP
  * ffmpeg HLS push (with mpegts)
  * ffmpeg HLS push (with mp4, as described above)
  * Standalone T with MP4 and mpegts
- Visual inspection of multipart hexdump (but did not attempt to parse + extract)
- Checked with GCS to verify file extension and content type


**Does this pull request close any open issues?**

Fixes https://github.com/livepeer/go-livepeer/issues/1418


**Checklist:**
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] README and other documentation updated
- [ ] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

j0sh committed:
net: Regenerate proto file with profile format.

j0sh committed:
server: Take MP4s as input for HTTP push

Most of the changes are in the following areas:

* Adds a format -> extension lookup function
* Adds a extension -> format inverse lookup function
* Adds a format -> mime type lookup function
* Protocol Buffers changes, and conversions to/from VideoProfile
* Slotting in the lookups and error handling where appropriate
* Lots and lots of tests.

The extensive testing and strict error handling is both to codify
intended behavior and to ensure misuse resistance to further changes.

j0sh committed:
core: Make the naming of video temp files explicit.

The extension doesn't matter, and may no longer be mpegts.

j0sh committed:
mod: Update LPMS to https://github.com/livepeer/lpms/commits/99fb05c90e8617ecfc3799b167d3de5600dedcf5

j0sh committed:
doc: Update push ingest docs with MP4 usage.

j0sh committed:
server: Set mime type for remote T responses.

darkdarkdragon reviewed (changes requested):
`mp4` muxer/demuxer should be added to `install_ffmpeg.sh`

j0sh commented:
@darkdarkdragon Ah yes.

Added a unit test that will iterate through all our formats, to catch this type of oversight in the future:  https://github.com/livepeer/go-livepeer/pull/1429/commits/8abe8abe03532820c8d8fe382d83f5220f8ddbc8 . That [failed](https://circleci.com/gh/livepeer/go-livepeer/5120?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) exactly  as expected.

Added mp4 to the muxer/demuxer and tests pass now. https://github.com/livepeer/go-livepeer/pull/1429/commits/06b87835cc5f671d342b60a91d821cc24759575f

j0sh commented:
Rebased to fix up a merge conflict. Git SHAs have changed but commit structure is otherwise the same in case someone was in the middle of reviewing things (eg, still contains the pending fixups from https://github.com/livepeer/go-livepeer/pull/1429#issuecomment-606856207)

j0sh committed:
server: Populate SegData.Profiles with bogus data.

This allows us to fail out quickly on orchestrators that do not
recognize the FullProfile field.

j0sh committed:
net, server: Add SegData.FullProfiles2 field.

Only use this field for non-mpegts profiles in order to maintain
compatibility with older orchestrators. If an older orchestrator
receives a request for non-mpegts transcoding, it will fail out.

j0sh commented:
https://github.com/livepeer/go-livepeer/pull/1429/commits/2c075d9ee24fb8f1b8e5f5eb9dabb78022355ec6 Updated to incorporate latest LPMS, including the libavfilter performance fix and MP4 support among other things:

https://github.com/livepeer/lpms/pull/176
https://github.com/livepeer/lpms/pull/177
https://github.com/livepeer/lpms/pull/178
https://github.com/livepeer/lpms/pull/179

darkdarkdragon reviewed (approved):
LGTM 🚢

yondonfu commented on server/segment_rpc.go:
```diff
@@ -133,13 +134,20 @@ func (h *lphttp) ServeSegment(w http.ResponseWriter, r *http.Request) {
+			break
+		}
+		name := fmt.Sprintf("%s/%d%s", segData.Profiles[i].Name, segData.Seq, ext)
+		// The use of := here is probably a bug?!?
```
Is the intent for any error encountered in this loop to be persisted via the `err` declared outside of this loop after we break out of so that the value of `err` can be used to populate the `net.TranscodeResult_Error` on L171 and passed to B?

```golang
var result net.TranscodeResult
if err != nil {
    glog.Errorf(...)
    result = net.TranscodeResult{Result: &net.TranscodeResult_Error{Error: err.Error()}}
}
```

If so, then yeah I think the short declaration on this line is indeed a bug since the `err` variable declared outside of this loop would still be nil when we break out of the loop and as a result the error from `res.OS.SaveData()` would not be passed to B.

Fine with addressing this separately from this PR though.

j0sh replied:
That was the behavior I was expecting, yeah - failure to save should be propagated immediately to the B as an error.

But it looks like this has been here for a long time, and it's actually codified in a [unit test](https://github.com/livepeer/go-livepeer/blob/9659e5de2b63933f20bd34775c748781033b5074/server/segment_rpc_test.go#L337-L401). So I'm uncertain if ignoring the error was meant to address some other case. Best I can think of is something like using the results that were successfully saved, and retrying the rest? But since that's not anything we support right now, I'm also OK correcting the behavior.

j0sh replied:
Bug fixed in https://github.com/livepeer/go-livepeer/pull/1429/commits/9aca289d6e123cdd3541493bc6ee388d36d111a6

Note that we [debit any fees](https://github.com/livepeer/go-livepeer/blob/9aca289d6e123cdd3541493bc6ee388d36d111a6/server/segment_rpc.go#L163-L164) that may have accumulated from an earlier iteration, but we don't return the accompanying pixel counts anymore. Not sure if that'll lead to a problem if the O is used again for subsequent segments (eg, if a refresh occurs soon thereafter). If it seems like this needs more work / thought, I'll drop the commit from this PR and we can address it separately.

yondonfu replied:
> Not sure if that'll lead to a problem if the O is used again for subsequent segments (eg, if a refresh occurs soon thereafter)

With the per-stream suspension list, the O wouldn't be reused for awhile so I think this should only be a problem if there are only a small # of available Os (so B can't cycle through other Os before reusing one that returned an error).

That being said, I lean towards debiting fees only if we're returning segments/pixel counts and not debiting fees if we're returning a transcode error to minimize accounting issues (also feels right to not debit fees from a B's balance if O is not returning segments that could actually be used by B). Since this would require moving the `DebitFees()` call and writing some additional test cases I'm cool with dropping the commit and addressing separately.

j0sh replied:
Reverted in 998619c86ce6ba027a18dddbd1931bc6344294d1 - will remove the commit from the tree when rebasing if that's a go

yondonfu replied:
Sounds good 👍

yondonfu reviewed (approved):
LGTM! Left one note about a comment about a possible bug in existing code. I'll leave it up to you whether to address it in this PR or separately (I noted that I'm fine with doing it separately).
